oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGCRecommendations.java

```diff
@@ -321,6 +328,32 @@ public void evaluate(VersionGCStats stats) {
         }
     }
 
+    @Nullable
+    private Revision getLastAliveRevision(Clock clock, Checkpoints checkpoints) {
```

Comment:
this duplicates quite a bit of code (the `Checkpoints.getOldestRevisionToKeep`).

what about alternatively introducing a boolean, eg `cleanupExpiredCheckpoints` to `getOldestRevisionToKeep` to have just 1 method? Then the current `getOldestRevisionToKeep()` could just call `getOldestRevisionToKeep(true)`, and the `VersionGCRecommendations` could call `getOldestRevisionToKeep(false)`?

Reply:
for example
```
        if (op.hasChanges()) {
```
in `getOldestRevisionToKeep` could be changed to
```
        if (cleanupExpiredCheckpoints && op.hasChanges()) {
```

Reply:
Ok, didn't do that because that method is used in 15 places, and thought it may be too many unnecessary changes.

Reply:
You can also keep the original method and have that invoke the new one with the default being `true`. That is probably good as changing 15 places seems problematic indeed

Reply:
Yes, did that already [in this commit](https://github.com/apache/jackrabbit-oak/pull/1556/commits/50b9eb6b12b876630923350d9d3b0caf56940d3e), it was easier than changing the 15 other places.